File: src/mvmctl/utils/yaml.py
"""
Typed field extraction helpers for YAML-parsed dictionaries.

These helpers validate and extract typed values from ``dict[str, Any]`` objects
produced by ``yaml.safe_load``.  They raise :class:`ValueError` on type or
presence failures so that callers can convert the error into whatever domain
exception is appropriate.
"""

from __future__ import annotations

from typing import Any


def require_str(data: dict[str, Any], key: str) -> str:
    """
    Return the value of *key* as a string, raising ``ValueError`` if absent or wrong-typed.

    Args:
        data: Mapping to read from.
        key: Field name.

    Returns:
        The string value.

    Raises:
        ValueError: If the key is missing or its value is not a :class:`str`.

    """
    value = data.get(key)
    if isinstance(value, str):
        return value
    raise ValueError(
        f"field '{key}' must be a string (got {type(value).__name__!r})"
    )


def optional_str(data: dict[str, Any], key: str) -> str | None:
    """
    Return the string value of *key*, or ``None`` if absent or non-string.

    Args:
        data: Mapping to read from.
        key: Field name.

    Returns:
        The string value, or ``None``.

    """
    value = data.get(key)
    return value if isinstance(value, str) else None


def optional_int(data: dict[str, Any], key: str) -> int | None:
    """
    Return the integer value of *key*, or ``None`` if absent or non-integer.

    Args:
        data: Mapping to read from.
        key: Field name.

    Returns:
        The integer value, or ``None``.

    """
    value = data.get(key)
    return value if isinstance(value, int) else None


def require_str_list(data: dict[str, Any], key: str) -> list[str]:
    """
    Return the value of *key* as a list of strings.

    An absent key is treated as an empty list.  A value that is a list but
    contains non-string items, or a value that is not a list at all, raises
    ``ValueError``.

    Args:
        data: Mapping to read from.
        key: Field name.

    Returns:
        List of strings (may be empty).

    Raises:
        ValueError: If the value is present but is not a list of strings.

    """
    value = data.get(key, [])
    if isinstance(value, list) and all(isinstance(item, str) for item in value):
        return list(value)
    raise ValueError(f"field '{key}' must be a list of strings")


def parse_set_val_list(data: dict[str, Any], key: str) -> list[tuple[str, str]]:
    """
    Return option/value pairs from *key* as a list of ``(option, value)`` tuples.

    Each entry in the YAML list may be either a ``{option: ..., value: ...}``
    mapping or a two-element sequence.  Any other shape raises ``ValueError``.

    An absent key is treated as an empty list.

    Args:
        data: Mapping to read from.
        key: Field name whose value is the list of option/value pairs.

    Returns:
        List of ``(option, value)`` tuples (may be empty).

    Raises:
        ValueError: If the value is not a list, or any entry has an unexpected shape.

    """
    items = data.get(key, [])
    if not isinstance(items, list):
        raise ValueError(f"field '{key}' must be a list")
    result: list[tuple[str, str]] = []
    for item in items:
        if isinstance(item, dict) and "option" in item and "value" in item:
            result.append((str(item["option"]), str(item["value"])))
        elif isinstance(item, (list, tuple)) and len(item) == 2:
            result.append((str(item[0]), str(item[1])))
        else:
            raise ValueError(
                f"field '{key}' entries must be {{option, value}} mappings or two-element lists"
            )
    return result

File: internal/infra/yaml.go
package infra

import (
	"fmt"
)

// RequireString extracts a required string field from a YAML map.
// Returns an error if the key is missing or the value is not a string.
func RequireString(m map[string]any, key string) (string, error) {
	v, ok := m[key]
	if !ok {
		return "", fmt.Errorf("field '%s' is required", key)
	}
	s, ok := v.(string)
	if !ok {
		return "", fmt.Errorf("field '%s' must be a string (got %T)", key, v)
	}
	return s, nil
}

// OptionalString returns a pointer to the string value of key,
// or nil if absent or not a string.
func OptionalString(data map[string]any, key string) *string {
	v, ok := data[key]
	if !ok {
		return nil
	}
	s, ok := v.(string)
	if !ok {
		return nil
	}
	return &s
}

// OptionalInt returns a pointer to the integer value of key,
// or nil if absent or not an integer (or bool).
//
// YAML true/false decode as bool in Go's yaml.v3, so we accept bool as valid
// int (true→1, false→0). float64 values (even round ones like 42.0) are
// rejected.
func OptionalInt(data map[string]any, key string) *int {
	v, ok := data[key]
	if !ok {
		return nil
	}
	switch n := v.(type) {
	case int:
		return &n
	case bool:
		if n {
			one := 1
			return &one
		}
		zero := 0
		return &zero
	default:
		return nil
	}
}

// RequireStrList returns the value of key as a list of strings.
// An absent key is treated as an empty list. Returns an error if the value
// is present but not a list of strings.
func RequireStrList(data map[string]any, key string) ([]string, error) {
	v, ok := data[key]
	if !ok {
		return []string{}, nil
	}
	list, ok := v.([]any)
	if !ok {
		return nil, fmt.Errorf("field '%s' must be a list of strings", key)
	}
	result := make([]string, 0, len(list))
	for _, item := range list {
		s, ok := item.(string)
		if !ok {
			return nil, fmt.Errorf("field '%s' must be a list of strings", key)
		}
		result = append(result, s)
	}
	return result, nil
}

// SetValEntry represents a parsed option/value pair from YAML.
type SetValEntry struct {
	Option string
	Value  string
}

// ParseSetValList parses option/value pairs from YAML data map under the given key.
// Each entry in the list should be either:
// - a map with "option" and "value" keys, or
// - a two-element list where item[0]=option, item[1]=value
func ParseSetValList(data map[string]any, key string) ([]SetValEntry, error) {
	v, ok := data[key]
	if !ok {
		return nil, nil
	}
	list, ok := v.([]any)
	if !ok {
		return nil, fmt.Errorf("field '%s' must be a list", key)
	}
	var result []SetValEntry
	for _, item := range list {
		switch entry := item.(type) {
		case map[string]any:
			option, ok := entry["option"].(string)
			if !ok {
				return nil, fmt.Errorf("field '%s' entries must have a string 'option' field", key)
			}
			value, ok := entry["value"].(string)
			if !ok {
				return nil, fmt.Errorf("field '%s' entries must have a string 'value' field", key)
			}
			result = append(result, SetValEntry{Option: option, Value: value})
		case []any:
			if len(entry) != 2 {
				return nil, fmt.Errorf("field '%s' list entries must have exactly 2 elements (option, value)", key)
			}
			option, ok := entry[0].(string)
			if !ok {
				return nil, fmt.Errorf("field '%s' list entry option must be a string", key)
			}
			value, ok := entry[1].(string)
			if !ok {
				return nil, fmt.Errorf("field '%s' list entry value must be a string", key)
			}
			result = append(result, SetValEntry{Option: option, Value: value})
		default:
			return nil, fmt.Errorf("field '%s' entries must be {option, value} mappings or two-element lists", key)
		}
	}
	return result, nil
}
